lumeer-core/src/main/java/io/lumeer/core/facade/SearchFacade.java

```diff
@@ -61,306 +65,258 @@
+   }
 
-      return new ArrayList<>(collections);
+   private SearchQuery buildSearchQuery(Query query) {
+      return SearchQuery.createBuilder(authenticatedUser.getCurrentUser().getEmail())
+                        .groups(authenticatedUserGroups.getCurrentUserGroups())
+                        .queryStems(query.getStems(), query.getFulltexts())
+                        .page(query.getPage())
+                        .pageSize(query.getPageSize())
+                        .build();
    }
 
-   public List<Collection> searchCollectionsByView(final View view, final boolean allRights) {
-      java.util.Collection<Collection> collections = getQueryCollections(view.getQuery(), view, false);
+   public List<Document> searchDocuments(final Query query) {
+      final List<Collection> collections = getReadCollections();
 
-      if (allRights) {
-         return new ArrayList<>(collections);
+      if (query.isEmpty()) {
+         return new ArrayList<>(getChildDocuments(searchDocumentsByEmptyQuery(query, collections)));
+      } else if (query.containsStems()) {
+         return new ArrayList<>(searchDocumentsByStems(query, collections));
       } else {
-         return collections.stream().map(this::mapResource).collect(Collectors.toList());
+         return new ArrayList<>(getChildDocuments(searchDocumentsByFulltexts(query, collections)));
       }
    }
 
-   private Set<Collection> getQueryCollections(final Query query) {
-      return getQueryCollections(query, permissionsChecker.getActiveView(), true).stream().map(this::mapResource).collect(Collectors.toSet());
-   }
-
-   private Set<Collection> getQueryCollections(final Query query, final View view, final boolean filterByDocs) {
-      Set<Collection> collections = new HashSet<>();
-
-      if (query.getLinkTypeIds() != null && query.getLinkTypeIds().size() > 0) {
-         final List<LinkType> linkTypes = query.getLinkTypeIds().stream().map(linkTypeDao::getLinkType).collect(Collectors.toList());
-         linkTypes.forEach(linkType -> collections.addAll(collectionDao.getCollectionsByIds(linkType.getCollectionIds())));
-      }
-
-      if ((query.getFulltext() != null && !query.getFulltext().isEmpty()) || (query.getFilters() != null && !query.getFilters().isEmpty())) {
-         collections.addAll(getCollectionsByDocumentsSearch(query, view, filterByDocs));
-      }
+   private List<Collection> getReadCollections() {
+      Set<Collection> userCollections = getUserCollections();
 
-      if (query.getDocumentIds() != null && !query.getDocumentIds().isEmpty()) {
-         collections.addAll(getCollectionsByDocumentsIds(query.getDocumentIds()));
-      }
-
-      if (collectionQueryIsNotEmpty(query) || isEmptyQuery(query)) {
-         collections.addAll(getCollectionsByCollectionSearch(query, view));
+      final View view = permissionsChecker.getActiveView();
+      if (view != null) {
+         userCollections.addAll(viewFacade.getViewsCollections(Collections.singletonList(view), true));
```

Comment:
Why cannot this be done by using directly `ViewDao`? We should try to minimize the number of dependencies between facades.

Reply:
Because this method use multiple daos to achieve task.